pkg/clusterapi/controlplane_test.go

```diff
@@ -0,0 +1,322 @@
+			},
+		},
+		{
+			name: "unstacked etcd",
+			controlPlane: &dockerControlPlane{
+				Cluster:                     capiCluster(),
+				ProviderCluster:             dockerCluster(),
+				KubeadmControlPlane:         kubeadmControlPlane(),
+				ControlPlaneMachineTemplate: dockerMachineTemplate(),
+				EtcdCluster:                 etcdCluster(),
+				EtcdMachineTemplate:         dockerMachineTemplate(),
+			},
+			want: []clusterapi.Object{
+				capiCluster(),
+				dockerCluster(),
+				kubeadmControlPlane(),
+				dockerMachineTemplate(),
+				etcdCluster(),
+				dockerMachineTemplate(),
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+			g.Expect(tt.controlPlane.Objects()).To(ConsistOf(tt.want))
+		})
+	}
+}
+
+func TestEnsureNewNameIfChangedObjectDoesNotExist(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	originalName := "my-machine-template-1"
+	mt := dockerMachineTemplate()
+	mt.Name = originalName
+	client := test.NewFakeKubeClient()
+
+	g.Expect(clusterapi.EnsureNewNameIfChanged(ctx, client, notFoundRetriever, withChangesCompare, mt)).To(Succeed())
+	g.Expect(mt.Name).To(Equal(originalName))
+}
+
+func TestEnsureNewNameIfChangedErrorReadingObject(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	mt := dockerMachineTemplate()
+	mt.Name = "my-machine-template"
+	client := test.NewFakeKubeClient()
+
+	g.Expect(
+		clusterapi.EnsureNewNameIfChanged(ctx, client, errorRetriever, withChangesCompare, mt),
+	).To(
+		MatchError(ContainSubstring("reading DockerMachineTemplate eksa-system/my-machine-template from API")),
+	)
+}
+
+func TestEnsureNewNameIfChangedErrorIncrementingName(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	mt := dockerMachineTemplate()
+	mt.Name = "my-machine-template"
+	client := test.NewFakeKubeClient()
+
+	g.Expect(
+		clusterapi.EnsureNewNameIfChanged(ctx, client, dummyRetriever, withChangesCompare, mt),
+	).To(
+		MatchError(ContainSubstring("incrementing name for DockerMachineTemplate eksa-system/my-machine-template")),
+	)
+}
+
+func TestEnsureNewNameIfChangedObjectNeedsNewName(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	mt := dockerMachineTemplate()
+	mt.Name = "my-machine-template-1"
+	client := test.NewFakeKubeClient()
+
+	g.Expect(clusterapi.EnsureNewNameIfChanged(ctx, client, dummyRetriever, withChangesCompare, mt)).To(Succeed())
+	g.Expect(mt.Name).To(Equal("my-machine-template-2"))
+}
+
+func TestEnsureNewNameIfChangedObjectHasNotChanged(t *testing.T) {
+	g := NewWithT(t)
+	ctx := context.Background()
+	originalName := "my-machine-template-1"
+	mt := dockerMachineTemplate()
+	mt.Name = originalName
+	client := test.NewFakeKubeClient()
+
+	g.Expect(clusterapi.EnsureNewNameIfChanged(ctx, client, dummyRetriever, noChangesCompare, mt)).To(Succeed())
+	g.Expect(mt.Name).To(Equal(originalName))
+}
+
+func TestControlPlaneUpdateImmutableObjectNamesNoKubeadmControlPlane(t *testing.T) {
```

Comment:
Since we aren't building any objects in the client, this test is basically a create cluster case, right? If KCP doesn't exists, `UpdateImmutableObjectNames` will return nil & everything should default to the default names